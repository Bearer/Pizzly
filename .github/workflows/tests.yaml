name: Run Unit & Integration Tests

on:
    push:
        branches:
            - master
            - staging/**
    pull_request:

concurrency:
    group: tests-${{ github.event.pull_request.number || github.ref }}
    cancel-in-progress: true

jobs:
    tests:
        runs-on: ubuntu-latest
        strategy:
            matrix:
                node-version: [18.x, 20.x]

        steps:
            - uses: actions/checkout@v3
              with:
                  fetch-depth: '0'

            - name: Use Node.js ${{ matrix.node-version }}
              uses: actions/setup-node@v4
              with:
                  cache: 'npm'
                  node-version: ${{ matrix.node-version }}
            - run: npm run build:hosted
            - run: npm run test:unit
            - run: npm run test:integration

            - name: Make sure flows.yaml is compiled
              run: |
                  node ./scripts/flows.js

                  REPO_DIRTY=$(git status --short)
                  if [[ ! -z "$REPO_DIRTY" ]]; then
                    git diff
                    echo ""
                    echo -e "\r⛔️ flows.yaml hasn't been compiled, please run:"
                    echo -e "node ./scripts/flows.js"
                    exit 1
                  fi
