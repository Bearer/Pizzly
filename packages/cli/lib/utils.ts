import https from 'https';
import axios from 'axios';
import type { NangoModel } from '@nangohq/shared';

export const configFile = 'nango.yaml';
export const NANGO_INTEGRATIONS_LOCATION = process.env['NANGO_INTEGRATIONS_LOCATION'] || './nango-integrations';

let hostport = process.env['NANGO_HOSTPORT'] || 'http://localhost:3003';
const cloudHost = 'https://api.nango.dev';
const stagingHost = 'https://nango-cloud-staging.onrender.com';

if (hostport.slice(-1) === '/') {
    hostport = hostport.slice(0, -1);
}

export function checkEnvVars() {
    if (hostport === 'http://localhost:3003') {
        console.log(`Assuming you are running Nango on localhost:3003 because you did not set the NANGO_HOSTPORT env var.\n\n`);
    } else if (hostport === cloudHost || hostport === stagingHost) {
        if (!process.env['NANGO_SECRET_KEY']) {
            console.log(`Assuming you are using Nango Cloud but your are lacking the NANGO_SECRET_KEY env var.`);
        } else if (hostport === cloudHost) {
            console.log(`Assuming you are using Nango Cloud (because you set the NANGO_HOSTPORT env var to https://api.nango.dev).`);
        } else if (hostport === stagingHost) {
            console.log(`Assuming you are using Nango Cloud (because you set the NANGO_HOSTPORT env var to https://api.staging.nango.dev).`);
        }
    } else {
        console.log(`Assuming you are self-hosting Nango (becauses you set the NANGO_HOSTPORT env var to ${hostport}).`);
    }
}

export async function getConnection(providerConfigKey: string, connectionId: string) {
    checkEnvVars();
    const url = hostport + `/connection/${connectionId}`;
    return await axios
        .get(url, { params: { provider_config_key: providerConfigKey }, headers: enrichHeaders(), httpsAgent: httpsAgent() })
        .then((res) => {
            return res.data;
        })
        .catch((err) => {
            console.log(`‚ùå ${err.response?.data.error || JSON.stringify(err)}`);
        });
}

export function enrichHeaders(headers: Record<string, string | number | boolean> = {}) {
    if ((process.env['NANGO_HOSTPORT'] === cloudHost || process.env['NANGO_HOSTPORT'] === stagingHost) && process.env['NANGO_SECRET_KEY']) {
        // For Nango Cloud (unified)
        headers['Authorization'] = 'Bearer ' + process.env['NANGO_SECRET_KEY'];
    } else if (process.env['NANGO_SECRET_KEY']) {
        // For Nango OSS
        headers['Authorization'] = 'Basic ' + Buffer.from(process.env['NANGO_SECRET_KEY'] + ':').toString('base64');
    }

    headers['Accept-Encoding'] = 'application/json';

    return headers;
}

export function httpsAgent() {
    return new https.Agent({
        rejectUnauthorized: false
    });
}

export function getFieldType(field: string | NangoModel): string {
    if (typeof field === 'string') {
        let tsType = '';
        switch (field) {
            case 'char':
                tsType = 'string';
                break;
            case 'integer':
                tsType = 'number';
                break;
        }
        return tsType;
    } else {
        const nestedFields = Object.keys(field)
            .map((fieldName: string) => `  ${fieldName}: ${getFieldType(field[fieldName] as string | NangoModel)};`)
            .join('\n');
        return `{\n${nestedFields}\n}`;
    }
}

export function buildInterfaces(models: NangoModel): (string | undefined)[] {
    const interfaceDefinitions = Object.keys(models).map((modelName: string) => {
        if (modelName.charAt(0) === '_') {
            return;
        }
        const fields = models[modelName] as NangoModel;
        const singularModelName = modelName.charAt(modelName.length - 1) === 's' ? modelName.slice(0, -1) : modelName;
        const interfaceName = `${singularModelName.charAt(0).toUpperCase()}${singularModelName.slice(1)}`;
        const fieldDefinitions = Object.keys(fields)
            .map((fieldName: string) => {
                const fieldModel = fields[fieldName] as string | NangoModel;
                const fieldType = getFieldType(fieldModel);
                return `  ${fieldName}: ${fieldType};`;
            })
            .join('\n');
        const interfaceDefinition = `export interface ${interfaceName} {\n${fieldDefinitions}\n}\n`;
        return interfaceDefinition;
    });

    return interfaceDefinitions;
}
