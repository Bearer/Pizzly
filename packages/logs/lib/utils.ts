import { customAlphabet } from 'nanoid';
import { z } from 'zod';

function zodErrorToString(issues: z.ZodIssue[]) {
    return issues
        .map((issue) => {
            return `${issue.path.join('')} (${issue.code} ${issue.message})`;
        })
        .join(', ');
}

export function initGlobalEnv() {
    const schema = z.object({
        NANGO_DATABASE_URL: z.string().url().optional(),
        SERVER_RUN_MODE: z.enum(['DOCKERIZED', '']).optional(),
        NANGO_DB_HOST: z.string().optional().default('localhost'),
        NANGO_DB_PORT: z.coerce.number().optional().default(5432),
        NANGO_DB_USER: z.string().optional().default('nango'),
        NANGO_DB_NAME: z.string().optional().default('nango'),
        NANGO_DB_PASSWORD: z.string().optional().default('nango'),
        NANGO_DB_SSL: z.enum(['true', 'false']).optional().default('false'),
        NODE_ENV: z.enum(['production', 'development', 'test']).default('development'),
        CI: z.enum(['true', 'false']).default('false'),
        VITEST: z.enum(['true', 'false']).default('false')
    });

    const res = schema.safeParse(process.env);
    if (!res.success) {
        throw new Error(`Missing or invalid env vars: ${zodErrorToString(res.error.issues)}`);
    }

    return res.data;
}

export function initLogsEnv() {
    const schema = z.object({
        NANGO_LOGS_ES_URL: z.string().url(),
        NANGO_LOGS_ES_USER: z.string(),
        NANGO_LOGS_ES_PWD: z.string()
    });

    const res = schema.safeParse(process.env);
    if (!res.success) {
        throw new Error(`Missing or invalid env vars: ${zodErrorToString(res.error.issues)}`);
    }

    return res.data;
}

/**
 * Nanoid without special char to use in URLs
 */
export const alphabet = '346789ABCDEFGHJKLMNPQRTUVWXYabcdefghijkmnpqrtwxyz';
export const minSize = 8;
export const maxSize = 20;
export const nanoid = customAlphabet(alphabet, maxSize);
