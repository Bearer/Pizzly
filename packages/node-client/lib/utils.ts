import type { ProxyConfiguration } from './types';

export const validateProxyConfiguration = (config: ProxyConfiguration) => {
    const requiredParams = ['endpoint', 'providerConfigKey', 'connectionId'];

    requiredParams.forEach((param) => {
        if (typeof config[param] === 'undefined') {
            throw new Error(`${param} is missing and is required to make a proxy call!`);
        }
    });
};
