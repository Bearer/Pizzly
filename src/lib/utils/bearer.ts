/**
 * Bearer.sh
 *
 * Initialize the Bearer agent if the environment key is provided.
 * Bearer will monitor and shield the Pizzly instance from APIs failure.
 * Learn more: https://www.bearer.sh/
 *
 * To get your BEARER_AGENT_KEY, create an account on www.bearer.sh
 * then heads to https://app.bearer.sh/settings/key
 */

import Bearer from '@bearer/node-agent'

if (process.env.BEARER_AGENT_KEY) {
  console.log('Initializing')
  Bearer.init({ secretKey: process.env.BEARER_AGENT_KEY })
}
