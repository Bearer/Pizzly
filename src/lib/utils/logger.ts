import winston from 'winston'
import expressWinston from 'express-winston'
const isDevelopment = process.env.NODE_ENV !== 'production'

const format = winston.format.combine(
  winston.format.colorize(),
  isDevelopment ? winston.format.simple() : winston.format.json()
)
const transports = [new winston.transports.Console()]
const meta = !isDevelopment
const msg = '{{res.statusCode}} {{req.method}} {{res.responseTime}}ms {{req.url}}'
const expressFormat = false

export const requestLogger = expressWinston.logger({
  transports,
  format,
  meta,
  msg,
  expressFormat,
  colorize: true // Color the text and status code, using the Express/morgan color palette (text: gray, status: default green, 3XX cyan, 4XX yellow, 5XX red).
})

export const errorLogger = expressWinston.errorLogger({
  transports,
  format,
  meta,
  msg
})
