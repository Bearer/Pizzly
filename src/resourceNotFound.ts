import { Request, Response } from 'express'
export default (req: Request, res: Response) => {
  const message = [
    'The requested URL',
    req.originalUrl,
    'could not be found on this server.',
    "Please make sure you've entered the right URL and try again.",
    'For further help, please see https://docs.bearer.sh/help-and-support'
  ].join(' ')
  if (!res.headersSent) {
    // console.log('resourceNotFound', message)
    res.status(404)
    if (req.accepts('html')) {
      res.header('Content-Type', 'text/html')
      res.render('page-not-found-error', {
        method: req.method,
        error: '404: Page Not Found',
        callbackUrl: req.originalUrl
      })
    } else {
      res.header('Content-Type', 'application/json')
      res.json({
        error: {
          message,
          code: 'NOT_FOUND'
        }
      })
    }
  }
}
