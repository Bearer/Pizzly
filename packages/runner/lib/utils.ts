import { getLogger, stringifyError } from '@nangohq/utils';

export const logger = getLogger('Runner');

export async function httpFetch({ method, url, data }: { method: string; url: string; data: string }): Promise<void> {
    try {
        const res = await fetch(url, {
            method: method,
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json'
            },
            body: data
        });
        if (res.status > 299) {
            logger.error(`Error (status=${res.status}) sending '${data}' to '${url}': ${JSON.stringify(await res.json())}`);
        }
    } catch (err) {
        logger.error(`Error sending '${data}' to '${url}': ${stringifyError(err)}`);
    }
}
