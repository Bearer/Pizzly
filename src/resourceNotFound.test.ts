import request from 'supertest'

import resourceNotFound from './resourceNotFound'
import { baseApp } from '../tests/utils'

describe('resourceNotFound middleware', () => {
  const app = baseApp()
  app.use(resourceNotFound)

  it('Returns 404 NOT_FOUND JSON response for non-HTML-accepting requests', async () => {
    const response = await request(app)
      .get('/qwerty')
      .set('Accept', 'application/json')
      .expect(404)
      .expect('Content-Type', /json/)

    expect(JSON.parse(response.text)).toMatchSnapshot()
  })

  it('Returns 404 NOT_FOUND HTML response for HTML-accepting requests', async () => {
    await request(app)
      .get('/qwerty')
      .set('Accept', 'text/html')
      .expect(404)
      .expect('Content-Type', /html/)
  })
})
