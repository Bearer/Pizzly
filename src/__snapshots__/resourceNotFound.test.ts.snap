// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resourceNotFound middleware Returns 404 NOT_FOUND JSON response for non-HTML-accepting requests 1`] = `
Object {
  "error": Object {
    "code": "NOT_FOUND",
    "message": "The requested URL /qwerty could not be found on this server. Please make sure you've entered the right URL and try again. For further help, please see https://docs.bearer.sh/help-and-support",
  },
}
`;
